src/utils/decorators/tooltip-decorator/tooltip-decorator.js

```diff
@@ -187,25 +172,68 @@ const TooltipDecorator = (ComposedComponent) => {
         if (this.state.isVisible) this.setState({ isVisible: false });
         return;
       }
-
       const alignment = this.state.tooltipAlign || this.props.tooltipAlign || 'center';
       const position = this.state.tooltipPosition || this.props.tooltipPosition || 'top';
       const shifts = calculatePosition(tooltip, target);
+      const isTooltipOffScreenRight = this.isTooltipOffScreenRight(position, alignment, tooltip.offsetWidth, shifts);
 
-      this.realignOffscreenTooltip(shifts, tooltip.offsetWidth);
+      if (isTooltipOffScreenRight) {
+        this.realignOffscreenTooltip(position, alignment);
+      }
 
       if (position === 'top' || position === 'bottom') {
-        topPosition = shifts.tooltipDistances[position];
+        topPosition = shifts.tooltipDistances[position] + this.getTooltipDistanceOffset(position);
         leftPosition = shifts.vertical[alignment];
       } else {
         topPosition = shifts.horizontal[alignment];
-        leftPosition = shifts.tooltipDistances[position];
+        leftPosition = shifts.tooltipDistances[position] + this.getTooltipDistanceOffset(position);
       }
 
       styleElement(tooltip, 'top', `${topPosition}px`);
       styleElement(tooltip, 'left', `${leftPosition}px`);
     };
 
+    isTooltipOffScreenRight = (position, alignment, tooltipWidth, shifts) => {
+      let pointerX;
+
+      if (position === 'right') {
+        pointerX = shifts.tooltipDistances.right;
+      } else {
+        pointerX = shifts.vertical[alignment];
+      }
+
+      return window.innerWidth < pointerX + tooltipWidth;
+    }
+
+    realignOffscreenTooltip = (position, alignment) => {
+      if (alignment === 'right' || position === 'left') return;
+
+      if (position === 'right') {
+        this.setState({ tooltipPosition: 'top', tooltipAlign: 'right' });
+      } else {
+        this.setState({ tooltipAlign: 'right' });
+      }
+    }
+
+    getTooltipDistanceOffset = (position) => {
+      if (!this.props.isThemeModern || !this.props.hasValidationTooltip || !this.props.size) return 0;
```

Comment:
isn't `size` a default property? perhaps checking for it here is redundant?

Reply:
it depends on the decorated component. Users could use a component that have no size property.